### DISCUSSION TITLE
Help updating graphical representation of physics bodies after physics simulation

### DISCUSSION THREAD

**Discussion by a1M322** [jrouwe/JoltPhysics#1382] [Q&A] [answered]
Hi first of all, thank you for your work! I've been trying Jolt Physics here and there and I'm facing a problem, although is more of a general question. It is related to my ignorance using physics engines as I'm fairly inexperience in 3D but im curious...

Right now I'm running a simulation on the entities of my world and after every physics update I go, look into a
 `hashmap<BodyIds, Entity>` and update the corresponding transform of the graphical representation of the entity. One of my problems is that not everything is of type Entity so I can not use this solution for every thing in the physics update. I've noticed there is also a `SetUserData(uint64 inUserData)` that can be used with bodies to probably indicate some index into some structure.
 
**Note**: I'm not using any ECS nor inheritance on my game logic.

**So my question is**: How does people normally handle this synchronization between the new transform of the body after the physics simulation and the update to the actual graphical representation of the physics body. If I had some ECS then using a phsyics component I could query this component for the entities that have them, but this is not the case, at least for know.
 
(Also, is there a Discord server?)

Thanks!

**Comment by jrouwe** [comment, marked as answer]
I would suggest to use `Body::Get/SetUserData`. You can store a 64-bit number, so that's plenty for a pointer to an `Entity` or perhaps something like:

```C++
class GameObjectInterface
{
    virtual void SetWorldTransform(...) = 0;
}

class Entity : public GameObjectInterface
{
    virtual void SetWorldTransform(...) override { ... }
}

class NotAnEntity : public GameObjectInterface
{
    virtual void SetWorldTransform(...) override { ... }
}
```

If that's not possible then at least the lower 3 bits of any pointer will be zero (allocations are usually 8 byte aligned on 32-bit platforms, 16 byte aligned on 64-bit platforms), so you could use those bits to store the type of object that you're pointing at.

After `PhysicsSystem::Update`:
* Loop over all `PhysicsSystem::GetActiveBodies` plus all bodies that were deactivated that update through `BodyActivationListener::OnBodyDeactivated`
* Cast `Body::GetUserData` to `GameObjectInterface` 
* Call `GameObjectInterface::SetWorldTransform` to update the transforms of the game objects.

> (Also, is there a Discord server?)

No there isn't. I think github discussions keep stuff better organized and searchable.
